synapse/storage/databases/main/stats.py

```diff
@@ -210,34 +210,56 @@ async def update_room_state(self, room_id: str, fields: Dict[str, Any]) -> None:
         * topic
         * avatar
         * canonical_alias
+        * guest_access
 
-        A is_federatable key can also be included with a boolean value.
+        An is_federatable key can also be included with a boolean value.
 
         Args:
             room_id: The room ID to update the state of.
             fields: The fields to update. This can include a partial list of the
                 above fields to only update some room information.
         """
-        # Ensure that the values to update are valid, they should be strings and
-        # not contain any null bytes.
+        # Ensure that the values to update are valid. str values should not
+        # contain any null bytes.
         #
         # Invalid data gets overwritten with null.
         #
         # Note that a missing value should not be overwritten (it keeps the
         # previous value).
-        sentinel = object()
-        for col in (
+        str_columns = [
             "join_rules",
             "history_visibility",
             "encryption",
             "name",
             "topic",
             "avatar",
             "canonical_alias",
-        ):
-            field = fields.get(col, sentinel)
-            if field is not sentinel and (not isinstance(field, str) or "\0" in field):
-                fields[col] = None
+            "guest_access",
+        ]
+        other_columns = [
+            "is_federatable",
+        ]
+
+        state_dict = {}
+        sentinel = object()
+
+        for column_name in fields.keys():
+            field = fields.get(column_name, sentinel)
+            if field is sentinel:
+                continue
+
+            if column_name in str_columns:
+                if not isinstance(field, str) or "\0" in field:
+                    # This field contained invalid str data, replace its contents with NULL
+                    state_dict[column_name] = None
+            elif column_name in other_columns:
+                if isinstance(field, str):
+                    # This field contained invalid data, replace its contents with NULL
+                    state_dict[column_name] = None
+            else:
+                # Calling code should only include keys in the above column lists. Otherwise
+                # we should never reach here
+                assert "Unknown column name '%s'" % (column_name,)
```

Comment:
This doesn't seem correct.

Reply:
Oh, I don't even add the value to `state_dict` in the correct case :man_facepalming: Is that what you meant?

Otherwise the assert here was supposed to alert a developer that they've put something in `fields` that `update_room_state` didn't know about, but perhaps that's not a necessary thing to raise?

Reply:
This was not enough info...

I don't think this will do anything since it is missing a boolean statement, e.g. this becomes `assert bool("...")` which resolves to `assert True`, which won't do anything.

Since the callers to this are internal I suspect we should raise a `ValueError` in this case.

Reply:
Works for me.